data-analytics/iot-nirvana/pipeline/src/main/java/com/google/cloud/demo/iot/nirvana/pipeline/JsonToTableRow.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.demo.iot.nirvana.pipeline;
+
+import com.google.api.services.bigquery.model.TableFieldSchema;
+import com.google.api.services.bigquery.model.TableRow;
+import com.google.api.services.bigquery.model.TableSchema;
+import com.google.cloud.demo.iot.nirvana.common.CityTemperature;
+import com.google.cloud.demo.iot.nirvana.common.Message;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.TimeZone;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.values.KV;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Transformation converting a message from JSON to TableRow format for writing into BigQuery
+ */
+public class JsonToTableRow extends DoFn<KV<Message, CityTemperature>, TableRow> {
+
+  private static final long serialVersionUID = 1462827258689031685L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(JsonToTableRow.class);
+
+  private static final TimeZone PARIS_TIME_ZONE = TimeZone.getTimeZone("Europe/Paris");
+  private static final SimpleDateFormat DF = createDateFormat("YYYY-MM-dd'T'HH:mm:ss.SSS");
+  private static final SimpleDateFormat DF_YEAR = createDateFormat("YYYY");
+  private static final SimpleDateFormat DF_MONTH = createDateFormat("MM");
+  private static final SimpleDateFormat DF_DAY = createDateFormat("dd");
+  private static final SimpleDateFormat DF_HOUR = createDateFormat("HH");
+  private static final SimpleDateFormat DF_MINUTE = createDateFormat("mm");
+  private static final SimpleDateFormat DF_SECOND = createDateFormat("ss");
+  private static final SimpleDateFormat DF_FRAME = createDateFormat("SSS");
+
+  @ProcessElement
+  public void processElement(ProcessContext c) {
+    // Read the next message to process
+    KV<Message, CityTemperature> temperatureTuple = c.element();
```

Comment:
Done in commit eaa8b523ac4f242338c9d6891013b4b1dde3b06d.
Added city metadata in the Message POJO so that a PCollection of Message is sent donwstream instead of a PCollection of KV. However, 1) the BigQuery table contains additional derivate columns calculated from the timestamp attribute and 2) the Message POJOs are also sent to the Transform for writing to Datastore, thus I don't see other option than coding the TableRow schema in the transform that writes to BigQuery.